### PR TITLE
[SPARK-21956][CORE] Fetch up to max bytes when buf really been released

### PR DESCRIPTION
**Pull request by caneGuy**
## What changes were proposed in this pull request?

Right now,ShuffleBlockFetcherIterator will decrease bytesInFlight immediatily when take from results queue.But currentresult's bytebuf has not been released at that time.Then direct memory may be a little out of control.
We should decrease bytesInFlight when currentresult's bytebuf really been released.

## How was this patch tested?

Exsisted Unit test

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala [hunk 1]
```diff
@@ -157,7 +157,14 @@ final class ShuffleBlockFetcherIterator(
   private[storage] def releaseCurrentResultBuffer(): Unit = {
     // Release the current buffer if necessary
     if (currentResult != null) {
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala]: I can't evaluate the change, but, the 'null' case is now already handled by your `case _`

**Comment by kiszk** [comment]
IIUC, this change defers sending requests to remote notes until `releaseCurrentResultBuffer()` will be called.
Do you have any benchmark results to confirm no performance degradation?

**Comment by caneGuy** [comment]
Yes you are right @kiszk Benchmark on the way.I will put here later.Thanks for your time @kiszk

**Comment by caneGuy** [comment]
Sorry for replying so late.
I add some benchmark testing for this pr @kiszk .
And @jerryshao could you help review this pr?Thanks
```
Running benchmark: Benchmark fetch before vs after releasing buffer
  Running case: Testing fetch before releasing!
  Stopped after 10 iterations, 2423 ms
  Running case: Testing fetch after releasing!
  Stopped after 18 iterations, 2036 ms

Java HotSpot(TM) 64-Bit Server VM 1.8.0_25-b17 on Linux 4.4.0-64-generic
Intel(R) Core(TM) i7-6700 CPU @ 3.40GHz
Benchmark fetch before vs after releasing buffer: Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------
Testing fetch before releasing!                 46 /  242        345.0           2.9       1.0X
Testing fetch after releasing!                  73 /  113        215.7           4.6       0.6X
```
```
Running benchmark: Benchmark fetch before vs after releasing buffer
  Running case: Testing fetch before releasing!
  Stopped after 10 iterations, 3888 ms
  Running case: Testing fetch after releasing!
  Stopped after 10 iterations, 3970 ms

Java HotSpot(TM) 64-Bit Server VM 1.8.0_25-b17 on Linux 4.4.0-64-generic
Intel(R) Core(TM) i7-6700 CPU @ 3.40GHz
Benchmark fetch before vs after releasing buffer: Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------
Testing fetch before releasing!                100 /  389        157.8           6.3       1.0X
Testing fetch after releasing!                 151 /  397        104.3           9.6       0.7X
```
```
Running benchmark: Benchmark fetch before vs after releasing buffer
  Running case: Testing fetch before releasing!
  Stopped after 15 iterations, 2016 ms
  Running case: Testing fetch after releasing!
  Stopped after 14 iterations, 2110 ms

Java HotSpot(TM) 64-Bit Server VM 1.8.0_25-b17 on Linux 4.4.0-64-generic
Intel(R) Core(TM) i7-6700 CPU @ 3.40GHz
Benchmark fetch before vs after releasing buffer: Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------
Testing fetch before releasing!                 43 /  134        363.8           2.7       1.0X
Testing fetch after releasing!                  99 /  151        158.1           6.3       0.4X
```

**Comment by caneGuy** [comment]
@cloud-fan  Sorry for bothering you.Could you help review this?Thanks

File: core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala [hunk 2]
```diff
@@ -439,6 +443,8 @@ final class ShuffleBlockFetcherIterator(
                   fetchRequests += FetchRequest(address, Array((blockId, size)))
                   result = null
                 }
+                bytesInFlight -= size
+                fetchUpToMaxBytes()
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala]: we may throw exception before hitting this branch

Comment (caneGuy) [reply core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala]: Thanks @cloud-fan 
IIUC, i think when  `fetchfailedException` thrown there is no need to consider this.Which is also the original design.

**Comment by cloud-fan** [comment]
Technically speaking, `maxBytesInFlight` controls how many bytes are being transferred in network, so we should decrease `bytesInFlight` once we get data from the result queue. The actual memory usage usually beyond `maxBytesInFlight`, because the network transfer needs memory, and the fetched data which is being processed also needs memory. This is not a big problem because the worst case is we consume `maxBytesInFlight * 2` memory.

**Comment by caneGuy** [comment]
Thanks @cloud-fan i will close this pr.

**Pull request closed without merging**